### PR TITLE
Session Management in Node.js

### PR DESCRIPTION
**Pull request by Catemacharia**
### Article checklist
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

### CODE REVIEW

**Comment by Catemacharia** [comment]
This fixes #2019

**Comment by hectorkambow** [comment]
plagiarism check done

File: articles/session-management-in-nodejs-using-expressjs-and-express-session/index.md [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+    secret: "thisismysecrctekeyfhrgfgrfrty84fwir767",
+    saveUninitialized:true,
+    cookie: { maxAge: oneDay },
+    resave: false 
+}));
+```
+
+- `secret` - a random unique string key used to authenticate a session. It is stored in an environment variable and can't be exposed to the public. The key usually is long in a production environment, and it is a randomly generated string.
+
+- `resave` - takes a Boolean value. When true, it sets the session to save event without recording any changes on the sent request. `saveUninitialized` goes hand in hand with `resave`. They tell the session middleware how to react to different events in the browser.
```
Comment (zolomohan) [inline articles/session-management-in-nodejs-using-expressjs-and-express-session/index.md:66]: The explanation here is hard to follow. I'm not 100% sure what you are trying to convey.

File: articles/session-management-in-nodejs-using-expressjs-and-express-session/index.md [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+app.get('/',(req,res) => {
+    session=req.session;
+    if(session.userid){
+        res.send("welcome User <a href=\'/logout'>click to logout</a>");
+    }else
+    res.sendFile('index.html',{root:__dirname})
+});
+
+app.post('/user',(req,res) => {
+    if(req.body.username == 'user1' && req.body.password== 'mypassword'){
```
Comment (zolomohan) [inline articles/session-management-in-nodejs-using-expressjs-and-express-session/index.md:184]: Can you store the username and password in variables? It'll be easier to understand and for readers to edit it.

**Review by zolomohan — CHANGES_REQUESTED** [review]
Hi @Catemacharia. I've some minor suggestions. Please go through them.

**Comment by zolomohan** [comment]
Hi @Catemacharia. Please ping me when you have updated the changes.

**Comment by Catemacharia** [comment]
Made changes @zolomohan

File: articles/session-management-in-nodejs-using-expressjs-and-express-session/index.md [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+    secret: "thisismysecrctekeyfhrgfgrfrty84fwir767",
+    saveUninitialized:true,
+    cookie: { maxAge: oneDay },
+    resave: false 
+}));
+```
+
+- `secret` - a random unique string key used to authenticate a session. It is stored in an environment variable and can't be exposed to the public. The key usually is long in a production environment, and it is a randomly generated string.
+
+- `resave` and `saveUninitialized` - tell the session middleware how to react to different events in the browser-based on the sent request.
```
Comment (zolomohan) [inline articles/session-management-in-nodejs-using-expressjs-and-express-session/index.md:64]: What events? Can you explain which exact events you are talking about and how the middleware reacts to it.

**Comment by Catemacharia** [comment]
Made changes @zolomohan

**Comment by Catemacharia** [comment]
@zolomohan, have you seen the updates?

File: articles/session-management-in-nodejs-using-expressjs-and-express-session/index.md [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+
+The server will send a cookie to the client's browser. The session ID is going to be placed inside this cookie.
+
+Once the client browser saves this cookie, it's then going to send that cookie along with each subsequent request to the server.
+
+The server is going to validate the cookie against the session ID. If the validation is successful, the user is granted access to the requested resources on the server.
+
+If the credentials are invalid, the server will not grant this user access to the resources. No session will be initialized, and no cookie will be saved.
+
+When the user decides to log out, the server will destroy (`req.session.destroy();`) the session and clear out the cookie on the client-side. Cookies are cleared in the browser when the `maxAge` expires.
```
Comment (zolomohan) [inline articles/session-management-in-nodejs-using-expressjs-and-express-session/index.md:229]: Hi @Catemacharia. This part of the tutorial is not well structured. 

You are giving the entire code at once and then talking about what the code does. But the reader will not know which part of the code is doing which step. 

Please split the code and explain how to build the JS file step-by-step instead of giving it all at once. Please take a look at other articles to better understand how to structure this. 

For example, take a look at [this article](https://www.section.io/engineering-education/text-to-speech-in-javascript/) that has been already published.

**Comment by Catemacharia** [comment]
@zolomohan, Check the newly added format.

**Comment by Catemacharia** [comment]
I have also updated the PR to match the new  EngEd Repo structure

**Comment by hectorkambow** [comment]
Thank you v much for your patience as we worked through our Repo re-structuring process. 

The purpose behind this restructure - was to better allow us to handle more volume and allow students to publish more articles and get more involved with the community (with open-source contribution and suggestions). 

With most of the changes coming to completion - we will be sure to get your article published asap. 

Thank you again - and we are looking forward to more amazing articles.

**Comment by Catemacharia** [comment]
> @zolomohan, Check the newly added format.

@zolomohan, AM still waiting for an update.

File: content/articles/session-management-in-nodejs-using-expressjs-and-express-session/index.md [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+}));
+```
+
+#### Parse the HTML form
+As we explained, the Body-parser will help us parser an HTTP POST method request from an HTML document. Set the sever to use the Body-parser so that we can access this POST method. We also need to serve the CSS styling to format the outlook of the HTML form.
+
+```js
+// get our app to use a body-parser
+
+app.use(bodyParser.urlencoded({ extended: true }))
```
Comment (zolomohan) [inline content/articles/session-management-in-nodejs-using-expressjs-and-express-session/index.md:187]: BodyParser is deprecated. Please use express methods to perform this operation. 
Remove it from the project dependencies throughout the article.

File: content/articles/session-management-in-nodejs-using-expressjs-and-express-session/index.md [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+- The server will create a temporary user session with a random string known as a session ID to identify that session.
+- The server will send a cookie to the client's browser. The session ID is going to be placed inside this cookie.
+
+Once the client browser saves this cookie, it will send that cookie along with each subsequent request to the server. The server will validate the cookie against the session ID. If the validation is successful, the user is granted access to the requested resources on the server.
+
+If the credentials are invalid, the server will not grant this user access to the resources. No session will be initialized, and no cookie will be saved.
+
+```js
+app.post('/user',(req,res) => {
+    if(req.body.username == myusename && req.body.password == mypassword){
```
Comment (zolomohan) [inline content/articles/session-management-in-nodejs-using-expressjs-and-express-session/index.md:245]:
```suggestion
    if(req.body.username == myusername && req.body.password == mypassword){
```

**Review by zolomohan — CHANGES_REQUESTED** [review]
Hi @Catemacharia. After logging in, When I reach the `/` home route, I don't see the logout button. Instead, I see the form again. Could you check why this is happening?

**Comment by Catemacharia** [comment]
It should work fine now @zolomohan

**Comment by Catemacharia** [comment]
> Hi @Catemacharia. After logging in, When I reach the `/` home route, I don't see the logout button. Instead, I see the form again. Could you check why this is happening?

@zolomohan I hope you got the newly added commit

File: content/articles/session-management-in-nodejs-using-expressjs-and-express-session/index.md [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+
+1. `http://localhost:4000/`
+
+This will render and serve the HTML form to the client to fill in the login credentials. If the user is logged in, we'll display a logout link.
+
+```js
+app.get('/',(req,res) => {
+    session=req.session;
+    if(session.userid){
+        res.send("welcome User <a href=\'/logout'>click to logout</a>");
```
Comment (zolomohan) [inline content/articles/session-management-in-nodejs-using-expressjs-and-express-session/index.md:223]:
```suggestion
        res.send("Welcome User <a href=\'/logout'>click to logout</a>");
```

**Comment by zolomohan** [comment]
Is this your first PR @Catemacharia?

**Comment by Catemacharia** [comment]
No, this is my second.

**Review by zolomohan — APPROVED** [review]

**Comment by hectorkambow** [comment]
Great work @Catemacharia 
Awesome tutorial to follow along - please make sure to do a git pull on your forked repo before your next article 👍
